### PR TITLE
513 availability links

### PR DESCRIPTION
**Pull request by thatandromeda**
## Status
**READY**

#### What does this PR do?
Makes the 'action' buttons on the full record view actually work (place hold, recall, ILL, request scan).

#### How can a reviewer manually see the effects of these changes?
Look at the PR build. Search for whatever (in books & media) and go to its full record page. Lo, there are action buttons! Click on them. They do stuff. However, there are not always action buttons! Because not all actions are appropriate for all objects.

#### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/DI-513
- https://mitlibraries.atlassian.net/browse/DI-549

#### Todo:
- [x] Tests
- [ ] Documentation
- [ ] Stakeholder approval

#### Requires Database Migrations?
NO

#### Includes new or updated dependencies?
NO

### CODE REVIEW

**Comment by thatandromeda** [comment]
Making a list to keep track of things that @frrrances may need to look at:
- [ ] Order of `@options` in `ButtonMaker`
- [ ] Fact that button HTML has been pushed down into `ButtonMaker`

**Comment by thatandromeda** [comment]
Most of the CodeClimate issues are roundabout ways of saying "circ rules are super complicated". Since this is an accurate reflection of ground truth, I have not tried to address these issues.

I don't know the rails-idiomatic place to put `ButtonMaker`; ideas welcome.

File: app/models/aleph_item.rb [hunk 1]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+# @options is a list of the names of each action that might be available for an
+# item.
+# For each action in options, the following functions must exist:
+# * eligible_for_#{action}? - returns true/false
+# * make_button_for_#{action} - returns HTML of relevant button
+# This should probably have related to the Great Not-Completed AlephObject
+# Refactor, but it just got too janky to deal with all the edge cases in our
+# circ logic - too hard to tell whether they'd all been dealt with, except by
+# pulling it all into a special-purpose, encapsulated place.
+class ButtonMaker
```
Comment (JPrevost) [inline app/models/aleph_item.rb]: We should probably put the ButtonMaker class in it's own file. It's pretty big to just ride along in the AlephItem file.

File: app/models/aleph_item.rb [hunk 2]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+    ].any?
+  end
+
+  def host_values
+    if Rails.env.production?
+      sfx_host = 'https://sfx.mit.edu/sfx_local'
+      request_host = 'library.mit.edu'
+    else
+      sfx_host = 'https://sfx.mit.edu/sfx_test'
+      request_host = 'walter.mit.edu'
```
Comment (JPrevost) [inline app/models/aleph_item.rb]: We should use prod SFX for all all environments.

File: app/models/aleph_item.rb [hunk 3]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+    end
+    [sfx_host, request_host]
+  end
+
+  def encoded_location
+    location = if @collection == 'Off Campus Collection'
+                 'OCC'
```
Comment (JPrevost) [inline app/models/aleph_item.rb]: Huh?

File: app/models/aleph_item.rb [hunk 4]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+    @z30status = @item.xpath('z30/z30-item-status').text
+    # Yes, this excludes `z30/` on purpose.
+    @z30status_code = @item.xpath('z30-item-status-code').text
+  end
+
+  def all_buttons
+    buttons = []
+    @options.each do |option|
+      eligibility_func = "eligible_for_#{option}?"
+      if method(eligibility_func).call
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply app/models/aleph_item.rb]: Yeah. I thought about having like a dict that mapped verbs (that can be performed on items) to handlers, but I decided I would hate that more. Or I guess there could be a world where ButtonMaker got subclassed for each verb, and each subclass defined its own `eligible?` and `make` functions, but I also would really like to not be writing this code any more.

File: app/models/aleph_item.rb [hunk 5]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+  # to show the `request scan` button. We don't want to disappoint users by
+  # offering them an option here that they won't be able to follow through on.
+  def eligible_for_scan?
+    [
+      call_number_valid_for_scan?,
+      z30status_valid_for_scan?,
+      collection_valid_for_scan?,
+      status_valid_for_scan?,
+      library_valid_for_scan?
+      # We should also exclude the subject header 'Standards, Engineering',
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply app/models/aleph_item.rb]: I just assumed "copyright?" and kept walking. Regardless I don't have subject headers available to me in this context (unless I want to yank them from EBSCO and insert them, a la the OCLC number, for just this case, but I do not).

File: app/models/aleph_item.rb [hunk 6]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+      # but we don't have subject headers in the Aleph data, so we aren't.
+    ].all?
+  end
+
+  # Some items are eligible for ILL (albeit only through BorrowDirect) even
+  # though they are also available in the library.
+  # Unfortunately we don't know if these items are actually *available* in BD.
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply app/models/aleph_item.rb]: 🦄

File: app/models/aleph_item.rb [hunk 7]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+
+  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Misc utilities ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  # Items with these process & item statuses may be put on hold/recalled.
+  def requestable?
+    return false if @on_reserve
+    [
+      # Items with these status codes may be requested from any library,
+      # except the Annex.
+      ['01', '03', '04', '05', '06', '07', '08', '09', '11', '12', '19', '20',
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply app/models/aleph_item.rb]: That's an awesome call. Will do.

File: app/models/aleph_item.rb (hunk 3 above)
Comment (thatandromeda) [reply app/models/aleph_item.rb]: I just do what I'm told.

File: app/models/aleph_item.rb (hunk 1 above)
Comment (thatandromeda) [reply app/models/aleph_item.rb]: Done.

File: app/models/aleph_item.rb [hunk 8]
```diff
@@ -72,3 +83,407 @@ def xml_status(id)
...
+    @options = ['call', 'contact', 'hold', 'ill', 'recall', 'scan', 'special_ill']
+
+    # Properties of items. This must go *after* setting @item and @oclc, but
+    # *before* setting @requestable.
+    set_item_properties
+
+    @requestable = requestable?
+  end
+
+  def set_item_properties
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply app/models/aleph_item.rb]: @🌈  
@🐰 
@🐱 
@🎆 
@🍻

File: app/models/aleph_item.rb (hunk 2 above)
Comment (thatandromeda) [reply app/models/aleph_item.rb]: Done.

File: test/models/button_maker_test.rb [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+  setup do
+    VCR.use_cassette('realtime aleph volume') do
+      @item = AlephItem.new.xml_status('MIT01001019412').xpath('//items').children.first
+    end
+    @oclc = '123456789'
+    @ButtonMaker = ButtonMaker.new(@item, @oclc)
+  end
+
+  test 'all buttons' do
+    puts @ButtonMaker.eligible_for_hold?
```
[the remark this answers is not in the document]
Comment (thatandromeda) [reply test/models/button_maker_test.rb]: Deleted.

File: app/models/aleph_item.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (JPrevost) [reply app/models/aleph_item.rb]: I'm pretty sure that works in rails fwiw

**Comment by JPrevost** [comment]
:shipit:

**Pull request merged**